userspace/libscap/engine/savefile/converter/converter.cpp

```diff
@@ -0,0 +1,404 @@
+	if(PPME_IS_ENTER(event_type)) {
+		return 'E';
+	} else {
+		return 'X';
+	}
+}
+
+static void clear_evt(uint64_t tid) {
+	if(evt_storage.find(tid) != evt_storage.end()) {
+		evt_storage[tid].reset();
+	}
+}
+
+static void store_evt(uint64_t tid, scap_evt *evt) {
+	// if there was a previous event for this tid, we can overwrite the pointer because it means we
+	// don't need it anymore. We need to keep the enter event until we retrieve it in the
+	// corresponding exit event, but if the same thread is doing another enter event it means the
+	// previous syscall is already completed.
+
+	clear_evt(tid);
+
+	scap_evt *tmp_evt = (scap_evt *)malloc(evt->len);
+	if(!tmp_evt) {
+		throw std::runtime_error("Cannot allocate memory for the enter event.");
+	}
+	memcpy(tmp_evt, evt, evt->len);
+	evt_storage[tid] = safe_scap_evt(tmp_evt);
+}
+
+static scap_evt *retrieve_evt(uint64_t tid) {
+	if(evt_storage.find(tid) != evt_storage.end()) {
+		return evt_storage[tid].get();
+	}
+	return nullptr;
+}
+
+static uint16_t get_param_len(scap_evt *evt, uint8_t num_param) {
+	if(evt->nparams <= num_param) {
+		std::string error = "Try to access len of param num '" + std::to_string(num_param) +
+		                    "' for event " + get_event_name((ppm_event_code)evt->type) + "_" +
+		                    get_direction_char((ppm_event_code)evt->type) +
+		                    " (num parameters: " + std::to_string(evt->type) + ").";
+		throw std::runtime_error(error);
+	}
+
+	// todo!: we need to manage LARGE_PAYLOAD events
+	uint16_t off_len = sizeof(scap_evt) + sizeof(uint16_t) * num_param;
+	uint16_t len = 0;
+	memcpy(&len, (char *)evt + off_len, sizeof(uint16_t));
+	return (uint32_t)len;
+}
+
+static char *get_param_ptr(scap_evt *evt, uint8_t num_param) {
+	if(evt->nparams <= num_param) {
+		std::string error = "Try to access param num '" + std::to_string(num_param) +
+		                    "' for event " + get_event_name((ppm_event_code)evt->type) + "_" +
+		                    get_direction_char((ppm_event_code)evt->type) +
+		                    " (num parameters: " + std::to_string(evt->type) + ").";
+		throw std::runtime_error(error);
+	}
+
+	char *ptr = (char *)evt + sizeof(scap_evt) + sizeof(uint16_t) * evt->nparams;
+	uint16_t ptr_off = 0;
+	for(auto i = 0; i < num_param; i++) {
+		uint16_t len = 0;
+		memcpy(&len, (char *)evt + sizeof(scap_evt) + sizeof(uint16_t) * i, sizeof(uint16_t));
+		ptr_off += len;
+	}
+
+	return ptr + ptr_off;
+}
+
+// This writes len + the param
+static void push_default_parameter(scap_evt *evt, uint16_t *params_offset, uint8_t param_num) {
+	// Please ensure that `new_evt->type` is already the final type you want to obtain.
+	// Otherwise we will access the wrong entry in the event table.
+	const struct ppm_event_info *event_info = &(g_event_info[evt->type]);
+	uint16_t len = scap_get_size_bytes_from_type(event_info->params[param_num].type);
+	char *ptr = scap_get_default_value_from_type(event_info->params[param_num].type);
```

Comment:
Is `scap_get_default_value_from_type` really needed? Can't we just use a 64bit ptr here?
We will copy only the first N bits, that will always be all 0s anyway.

Eg:
```C
uint64_t val = 0;
memcpy((char *)evt + *params_offset, (char *)&val, len);
```

Reply:
Great catch, thank you!